File: java/src/main/java/com/adsbynimbus/openrtb/user/Geo.java
package com.adsbynimbus.openrtb.user;

/**
 * This object encapsulates various methods for specifying a geographic location. When subordinate to a
 * Device object, it indicates the location of the device which can also be interpreted as the user’s current
 * location. When subordinate to a User object, it indicates the location of the user’s home base (i.e., not
 * necessarily their current location).
 * The lat/lon attributes should only be passed if they conform to the accuracy depicted in the type
 * attribute. For example, the centroid of a geographic region such as postal code should not be passed.
 */
public interface Geo {

    String LATITUDE = "lat"; // Float
    String LONGITUDE = "lon"; // Float (optional)
    String TYPE = "type"; // Integer
    String ACCURACY = "accuracy"; // Integer
    String COUNTRY = "country";
    String CITY = "city";

    // Location type
    int GPS = 1;
    int IP_LOOKUP = 2;
    int USER_PROVIDED = 3;

    /**
     * Builder for constructing a {@link Geo} object
     */
    interface Builder {

        /**
         * Set the latitude and accuracy
         *
         * @param latitude
         * @return this builder instance
         */
        Builder forLatitude(float latitude);

        /**
         * Set the country
         *
         * @param countryCode
         * @return this builder instance
         */
        Builder forCountry(String countryCode);

        /**
         * Set the longitude
         *
         * @param longitude
         * @return this builder instance
         */
        public Builder withLongitude(float longitude);

        /**
         * Set the accuracy of the location data as reported by the device
         *
         * @param accuracy - int
         * @return this builder instance
         */
        Builder withAccuracy(int accuracy);

        /**
         * Set the location type
         *
         * @param locationType
         * @return this builder instance
         */
        Builder withLocationType(int locationType);

        /**
         * Set the city
         *
         * @param city - {@link String}
         * @return this builder instance
         */
        Builder withCity(String city);
    }

    /**
     * Definition of {@link Geo} with all public mutable fields
     */
    class MutableGeo implements Geo {
        public Float lat;
        public Float lon;
        public Integer type;
        public Integer accuracy;
        public String country;
        public String city;
    }
}

File: go/request/rtb_twofive/geo.go
package twofive

import "github.com/francoispqt/gojay"

// Geo object encapsulates various methods for specifying a geographic location. When subordinate to a
// Device object, it indicates the location of the device which can also be interpreted as the user’s current
// location. When subordinate to a User object, it indicates the location of the user’s home base (i.e., not
// necessarily their current location).
type Geo struct {
	Lat       float64 `json:"lat,omitempty"       valid:"required"`
	Lon       float64 `json:"lon,omitempty"       valid:"required"`
	Type      int     `json:"type,omitempty"      valid:"range(1|3),optional"`    // 1,2,3 -> GPS/Location Services, IP Address, User provided (e.g., registration data)
	IPService int     `json:"ipservice,omitempty" valid:"range(1|4),optional"`    // 1,2,3,4 -> ip2location, Neustar (Quova), MaxMind, NetAcuity (Digital Element)
	Country   string  `json:"country,omitempty"   valid:"ISO3166Alpha3,optional"` // alpha 3
	City      string  `json:"city,omitempty"      valid:"optional"`
	Metro     string  `json:"metro,omitempty"     valid:"optional"`
	State     string  `json:"state,omitempty"     valid:"optional"`
}

// MarshalJSONObject implements MarshalerJSONObject
func (g *Geo) MarshalJSONObject(enc *gojay.Encoder) {
	enc.Float64KeyOmitEmpty("lat", g.Lat)
	enc.Float64KeyOmitEmpty("lon", g.Lon)
	enc.IntKeyOmitEmpty("type", g.Type)
	enc.IntKeyOmitEmpty("ipservice", g.IPService)
	enc.StringKeyOmitEmpty("country", g.Country)
	enc.StringKeyOmitEmpty("city", g.City)
	enc.StringKeyOmitEmpty("metro", g.Metro)
	enc.StringKeyOmitEmpty("state", g.State)
}

// IsNil checks if instance is nil
func (g *Geo) IsNil() bool {
	return g == nil
}

// UnmarshalJSONObject implements gojay's UnmarshalerJSONObject
func (g *Geo) UnmarshalJSONObject(dec *gojay.Decoder, k string) error {

	switch k {
	case "lat":
		return dec.Float64(&g.Lat)

	case "lon":
		return dec.Float64(&g.Lon)

	case "type":
		return dec.Int(&g.Type)

	case "ipservice":
		return dec.Int(&g.IPService)

	case "country":
		return dec.String(&g.Country)

	case "city":
		return dec.String(&g.City)

	case "metro":
		return dec.String(&g.Metro)

	case "state":
		return dec.String(&g.State)

	}
	return nil
}

// NKeys returns the number of keys to unmarshal
func (g *Geo) NKeys() int { return 0 }
